massa-module-cache/src/hd_cache.rs

```diff
@@ -0,0 +1,300 @@
+    /// # Arguments
+    /// * `hash`: hash associated to the module for which we want to set the cost
+    /// * `init_cost`: the new cost associated to the module
+    pub fn set_init_cost(&self, hash: Hash, init_cost: u64) {
+        let mut ser_metadata = Vec::new();
+        self.meta_ser
+            .serialize(&ModuleMetadata::Delta(init_cost), &mut ser_metadata)
+            .expect(DATA_SER_ERROR);
+        self.db
+            .put(metadata_key!(hash), ser_metadata)
+            .expect(CRUD_ERROR);
+    }
+
+    /// Sets a given module as invalid
+    pub fn set_invalid(&self, hash: Hash) {
+        let mut ser_metadata = Vec::new();
+        self.meta_ser
+            .serialize(&ModuleMetadata::Invalid, &mut ser_metadata)
+            .expect(DATA_SER_ERROR);
+        self.db
+            .put(metadata_key!(hash), ser_metadata)
+            .expect(CRUD_ERROR);
+    }
+
+    /// Retrieve a module
+    pub fn get(&self, hash: Hash, limit: u64, gas_costs: GasCosts) -> Option<ModuleInfo> {
+        let mut iterator = self
+            .db
+            .iterator(IteratorMode::From(&module_key!(hash), Direction::Forward));
+
+        if let (
+            Some(Ok((key_1, ser_module))),
+            Some(Ok((key_2, ser_metadata))),
+        ) = (iterator.next(), iterator.next())
+            && *key_1 == module_key!(hash)
+            && *key_2 == metadata_key!(hash)
+        {
+            let module = RuntimeModule::deserialize(&ser_module, limit, gas_costs).expect(MOD_DESER_ERROR);
+            let (_, metadata) = self
+                .meta_deser
+                .deserialize::<DeserializeError>(&ser_metadata)
+                .expect(DATA_DESER_ERROR);
+            let result = match metadata {
+                ModuleMetadata::Invalid => ModuleInfo::Invalid,
+                ModuleMetadata::NotExecuted => ModuleInfo::Module(module),
+                ModuleMetadata::Delta(delta) => ModuleInfo::ModuleAndDelta((module, delta)),
+            };
+            Some(result)
+        } else {
+            None
+        }
+    }
+
+    /// Try to remove as much as self.amount_to_snip entries from the db
+    fn snip(&mut self) {
+        let mut iter = self.db.raw_iterator();
+        let mut batch = WriteBatch::default();
+        let mut snipped_count: usize = 0;
+
+        while snipped_count < self.snip_amount {
+            // generate a random key
+            let mut rbytes = [0u8; 16];
+            rand::thread_rng().fill_bytes(&mut rbytes);
+            let key = *Hash::compute_from(&rbytes).to_bytes();
+
+            // take the upper existing key
+            iter.seek_for_prev(key);
+
+            // check iterator validity
+            if !iter.valid() {
+                continue;
+            }
+
+            // unwrap justified by above conditional statement
+            // read data key first because of seek_for_prev
+            let data_key = iter.key().unwrap();
+            batch.delete(data_key);
+            iter.prev();
```

Comment:
As you already take the first key that match your prefix shouldn't you go to the next to get the metadata key instead of prev one?

Reply:
I literally put a comment about this bro, but now that I see it maybe using `data` is misleading. We find the key from the randomly generated one using `seek_for_prev` so the first key we'll get will always be the metadata (aka `data_key`) after which we get the `module_key`.

Reply:
Resolved on the whiteboard for me